hoot-rnd/src/main/cpp/hoot/rnd/conflate/frechet/FrechetDistance.h

```diff
@@ -0,0 +1,168 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * --------------------------------------------------------------------
+ *
+ * The following copyright notices are generated automatically. If you
+ * have a new notice to add, please use the format:
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2016 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+#ifndef FRECHETDISTANCE_H
+#define FRECHETDISTANCE_H
+
+//  Boost
+#include <boost/multi_array.hpp>
+
+// Hoot
+#include <hoot/core/OsmMap.h>
+#include <hoot/core/Units.h>
+#include <hoot/core/algorithms/linearreference/WayLocation.h>
+#include <hoot/core/elements/Way.h>
+
+// TGS
+#include <tgs/SharedPtr.h>
+
+namespace hoot
+{
+
+typedef boost::multi_array<Meters, 2> frechet_matrix;
+typedef frechet_matrix::index frechet_index;
+typedef std::pair<int, int> vertex_match;
+typedef std::vector<vertex_match> subline_entry;
+typedef std::pair<Meters, subline_entry> frechet_subline;
+
+typedef shared_ptr<LineString> LineStringPtr;
+
+/** Class for calculating Frechet Distance between two ways and calculating maximal subline matches.
+ *  Algorithm developed from "A new merging process for data integration base on the descrete Frechet distance"
+ *  (http://thomas.devogele.free.fr/articles/avant_2003_selection/DevogeleSDH2002.pdf)
+ */
+class FrechetDistance
+{
+public:
+  /**
+   * @brief FrechetDistance class constructor
+   * @param map - pointer to the OSM map containing the two ways to be compared
+   * @param way1 - pointer to the first way to be compared
+   * @param way2 - pointer to the second way to be compared
+   */
+  FrechetDistance(const ConstOsmMapPtr& map, const ConstWayPtr& way1, const ConstWayPtr& way2, Radians maxAngle = M_PI);
+
+  /**
+   * @brief distance - calculate the Frechet distance between the two ways, _w1 and _w2
+   * @return Frechet distance in meters
+   */
+  Meters distance();
+
+  /**
+   * @brief maxSubline - find the maximum length, in nodes, subline for the two ways
+   * @param maxDistance - max distance between the two ways to consider as a match
+   * @return set of matching node indices that represent the subline
+   */
+  frechet_subline maxSubline(Meters maxDistance = 15.0);
+
+  /**
+   * @brief matchingSublines - find the set of non-overlapping maximal subline matches for the two ways
+   * @param maxDistance - max distance between the two ways to consider as a match
+   * @return set of matching non-overlapping subline matches
+   */
+  std::vector<frechet_subline> matchingSublines(Meters maxDistance = 15.0);
+
+protected:
+  /**
+   * @brief calculateMatrix
+   * @return distinct Frechet matrix of distances calculated
+   */
+  frechet_matrix calculateMatrix();
+  /**
+   * @brief advanceAndCheck
```

Comment:
That one slipped through.  I also added descriptions to the other functions in there too.